core-bundle/src/InsertTag/Flag/PhpFunctionFlag.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\InsertTag\Flag;
+
+use Contao\CoreBundle\DependencyInjection\Attribute\AsInsertTagFlag;
+use Contao\CoreBundle\InsertTag\InsertTagFlag;
+use Contao\CoreBundle\InsertTag\InsertTagResult;
+
+#[AsInsertTagFlag('addslashes')]
+#[AsInsertTagFlag('strtolower')]
+#[AsInsertTagFlag('strtoupper')]
+#[AsInsertTagFlag('ucfirst')]
+#[AsInsertTagFlag('lcfirst')]
+#[AsInsertTagFlag('ucwords')]
+#[AsInsertTagFlag('trim')]
+#[AsInsertTagFlag('rtrim')]
+#[AsInsertTagFlag('ltrim')]
+#[AsInsertTagFlag('urlencode')]
+#[AsInsertTagFlag('rawurlencode')]
+class PhpFunctionFlag implements InsertTagFlagInterface
+{
+    public function __invoke(InsertTagFlag $flag, InsertTagResult $result): InsertTagResult
+    {
+        static $allowedNames = null;
+
+        if (null === $allowedNames) {
+            foreach ((new \ReflectionClass(__CLASS__))->getAttributes(AsInsertTagFlag::class) as $attribute) {
+                $allowedNames[] = $attribute->newInstance()->name;
+            }
+        }
+
+        // Do not allow arbitrary PHP functions for security reasons
+        if (!\in_array($flag->getName(), $allowedNames, true)) {
+            throw new \LogicException(sprintf('Invalid flag "%s".', $flag->getName()));
+        }
+
+        return $result->withValue($flag->getName()($result->getValue()));
```

Comment:
IMHO most of these functions do not preserve the output type. E.g. `urlencode('body { color:red }')`  does not output valid CSS anymore.  `trim`, `ltrim`, `rtrim` are maybe the exception, for the others it's hard to tell IMHO. I think I would always set `OutputType::text` here as well.

On that note: Should we maybe make the output type a mandatory parameter when calling `InsertTagResult#withValue()`? This way it would be harder to miss setting the type. I think in nearly all cases the output type won't be preserved.

Thinking about this, this might be a small design issue: right now, we cannot really add more output types later on, because every implementer needs to know how to handle all of them. IMHO when registering an insert tag, you should probably need to specify the types your implementation *is* capable to handle. When overwriting an insert tag, you can then enhance this list. I think I would be fine if you could also tell the system that you can handle any type (as it virtually is right now) - but then you can never assume that the output type is kept (and will be kept for any type added in the future).

Reply:
(The last paragraph basically is true for any insert tag (flag) implementation. I won't comment it at the other places, though. :slightly_smiling_face:)

Reply:
> right now, we cannot really add more output types later on, because every implementer needs to know how to handle all of them.

I think they should handle unknown types by default like `text` and if they don’t know the source type they should output `text`.

> Should we maybe make the output type a mandatory parameter when calling `InsertTagResult#withValue()`?

I’ll fix the output type for addslashes, urlencode and rawurlencode. Making the type mandatory would just result in everyone copy/pasting `->withValue($value, $result->getOutputType())` so the added security would be minimal while making the API design worse IMO.

Reply:
Fixed the addslashes, urlencode and rawurlencode in #6195